ReadMe.md
![](https://bitbucket.org/uoyaeg/aegnec2/wiki/aegnec2.jpg )

# AEG NEC2: An open source MoM solver for electromagnetic simulations

The Applied Electromagnetics Group ([AEG][]) boundary element method ([BEM][]) 
method-of-moments ([MoM][]) solver, `aegnec2`, is an [Open Source][] derivative of 
the Numerical Electromagnetics Code Version 2 ([NEC2][]). It was developed in the in 
the [Department of Electronics][] at the [University of York][] for research in 
electromagnetic compatibility ([EMC][]) and computational electromagnetics ([CEM][]).


## Code Features

`aegnec2` has a number of enhancements compared to the original version of ([NEC2][]) including:

* Dynamically allocated memory using command line switches.

* Ability to use [BLAS][] based LU solvers from [LAPACK][] for increased
  performance if available.

* Fully double precision including all embedded constants.

* Improved user interface.

* Regression test suite.

* The `somnec` lossy ground model can be built directly into the nec
  binary making lossy ground calculations more straightforward and
  more flexible.

* Bug fixes and portability improvements.


## Requirements

### CMake (mandatory)

In order to compile and install the code from source using the build system provided a 
recent version of [CMake][] is required.

### Fortran77 compiler (mandatory)

The core code is written in [Fortran77][] and requires a compiler with the 
following [Fortran77][] extensions:

* Preprocessing of `.F` files either natively or transparently via `cpp`.

* `INCLUDE` statement to include files.

* `REAL*8` and `COMPLEX*16` data types.

* The following intrinsic functions: `DREAL`, `DIMAG`, `DCMPLX`, `DCONJG`,
  `CDABS`, `CDSQRT`, `CDEXP` and `DFLOAT`.

* List directed I/O from internal files.

* Portable equivalencing of `INTEGER` and `REAL*8`. Explicitly
  this must work:

      INTEGER I1 , I2
      REAL*8  R1

      COMMON /FRED/ I1 ,  I2

      EQUIVALENCE (I1,R1)

  such that R1 occupies the same storage as two INTEGER's. This
  restriction is likely to removed in a later version.

Timing information requires that the Fortran compiler supports one of
the following intrinsics: `ETIME`, `CPU_TIME` or `SECOND`. Otherwise all times 
will be reported as zero.

### ANSI C compiler (mandatory)
 
An ANSI [C89][] compiler is required for the front end. The C to
Fortran interface uses CMakes built-in support. It requires that the C data
types and Fortran data types map like:

C type   | Fortran type
:--------|:-------------
`int`    | `INTEGER`
`double` | `REAL*8`

It also requires that the length of character strings is passed by value 
as an `unsigned long` at the end of the argument list.

### BLAS libraries (optional)

A level 3 [BLAS][] library is required to use the [LAPACK][] based LU
solver. The reference implementation is available from the netlib
archives [BLAS][]. Many OS and compiler vendors supply highly optimised
versions of the BLAS libraries. Optimised BLAS/LAPACK libraries that 
have been tested with aegnec2 include:

* The portable [ATLAS][] library which provides a code generator for
  creating optimised BLAS (and some LAPACK) routines on a variety of
  platforms.
 
* The Intel Math Kernel Library ([MKL][]) for GNU/Linux.


## Installation

Installation instructions are contained in the file [Install.md][] in the 
source distribution.


## Documentation

Basic usage information for the `aegnec2` and `aegsomnec2` programs is 
available from the executables using the `-h` or `--help` flags, e.g.

    $ aegnec2 -h

or from the man pages

    $ man aegnec2
    
after installation . The input file follows the same syntax as
"standard" NEC2 except that full line and end of line comments are
allowed using a hash (`#`). This is useful for temporarily commenting
out a few cards. The input file name must end with a `.nec`
suffix. This may be relaxed in a later version. The suffix can be
omitted when running `aegnec2`. On-line [NEC2 manuals][] are available.

If `aegcnec2` is run with no options the default behaviour is similar to a
300 unknown version of "standard" NEC2. If you job has more than 300
segments you will need to specify the number with the -s option, for
example, if your job has 2000 segments then

    $ aegnec2 -s 2000 job.nec

will run NEC2 with sufficient storage for 2000 segments. You can
specify a FEW more segments than you actually need, however, it is
anti-social to run jobs which hog lots of memory for no purpose.

By default the core size of your job will be matched to exactly fit
the number of unknowns you specify. It is recommended that you run
`aegnec2` in this way.  However for very large jobs running on machines
with limited memory or many users it may be beneficial to limit the
core size. This can be done with the `-c` option. For example to run a
2000 segment job, which normally would use 64Mb of core storage, with
only 48Mb of core use

    $ aegnec2 -s 2000 -c 48m job.nec

This will use an out of core solver which is generally slower than an
in core solution but may be faster if the job would otherwise cause
your machine to use significant swap space. The suffices b, k and m
are used to denote bytes, kilobytes (1024 bytes) and megabytes (1024
kilobytes) respectively.

By default all the results are written into a file call job.res
(assuming the input file is called `job.nec`). If you use the `-M`
(`--old-multi-file`) option, e.g.

    $ aegnec2 -M job.nec

then any near field and far field data will also be written to
separate files called

File      | Contents
:---------|:-----------------------------------
`job.nre` | Near electric field data (NE card)
`job.nrh` | Near magnetic field data (NH card)
`job.rdp` | Radiation patterns (RP card)

for easier plotting. Impedance data will also be written to job.imp.
If multiple NE, NH or RP cards are used ALL the data is written to the
same file in the order in which it is calculated. If you would prefer
each NE, NH and RP card to put it's data in a separate file then use
the -m option (`--multi-file`):

    $ aegnec2 -m job.nec

In this case a separate file is created each time a card is used,
e.g. if four NE cards are used (or one NE with four frequency points):

    job-0001.nre
    job-0002.nre
    job-0003.nre
    job-0004.nre

will be created.


## Bugs and support

Known significant bugs are listed in the file doc/[Bugs.md][] in the source code. 

Please report bugs using the bitbucket issue tracker at
<https://bitbucket.org/uoyaeg/aegnec2/issues> or by email to <ian.flintoft@googlemail.com>.

For general guidance on how to write a good bug report see, for example:

* <http://www.chiark.greenend.org.uk/~sgtatham/bugs.html>
* <http://noverse.com/blog/2012/06/how-to-write-a-good-bug-report>

doc/News.md

# AEG NEC2 User Visible Changes

This file notes major user visible changes to the code. Details may be
found in the ChangeLog file.

## 15/05/2009 Version 0.8.0 Released

* Initial port to a new development system: 

  + CNEC2 is CMake now configured and built using CMake. Version 2.6.4 
    or later is required.

  + The source code is now managed using the Mercurial VCS.

* The license has been changed from version 2 to version 3 of the GNU 
  General Public License.

* The mini-LAPACK library has been update from version 2.0 to version
  3.2.1.

## 25/05/2005 Version 0.7.2 Released

* Somnec is now hopefully fully fixed with respect to saved 
  variables and initialisation.

* When multi-file output is used the impedance file is flushed
  to disk after every frequency step.

## 16/07/2002 Version 0.7.1 Released

* Renamed package from nec2d to cnec2. Also the executables change
  from nec2d to cnec2 and fnec2d becomes fnec2. This was done because
  there is already a version of NEC2 out there called nec2d!

* The gnu and testsuite directory have been renamed lib and tests
  repsectively.

* SOMNEC can now be built into the cnec2 executable using the
  --enable-builtin-somnec configure option. At a later date this will 
  become the default behaviour.

* --with-mingw configure option removed. use FFLAGS instead.

* Development now based on autoconf 2.53 and automake 1.6.2.

## 26/07/2001 Version 0.7.0 Released

* Documentation updates.

* New somnec C front end. The old interface is still available in the
  executable fsomnec.

* The 2000 unknown version of nec with Fortran front end is now
  called fnec2d (was called nec2000).

* Development now requires autoconf 2.50 and automake 2.4-p5.

## 04/05/2001 Version 0.6.8 Released

* Maintenance release.

* Fixed problems with jobname extraction with Fortran front end and
  problems with file opening on Linux systems using UNKNOWN
  status. All new files now opened using 'NEW'. This requires the old
  files are deleted by the C front end before calling the Fortran
  code.

## 12/01/2001 Version 0.6.7 Released

* Maintenance release. Updates to configuration and utility code.

* Fixed feature with filename dependent results in C - Fortran
  interface.

* Output files are now opened with status 'NEW' and the old output 
  files removed in the C front end. The Fortran front end will now 
  fail if there are existing output files.

## 13/10/2000 Version 0.6.6 Released

* Maintenance release. Updates to configuration and utility code.

## 04/10/2000 Version 0.6.5 Released

* Added support for using an external LAPACK library to the configure
  script. This is useful when using ATLAS for example.

* Now unquestionable licensed under the GPL.

* Development now requires CVS versions of automake (1.4a) and
 autoconf (2.49b).

* Bug fixes.

## 17/04/2000 Version 0.6.4 Released

* Bug fixes.

## 21/03/2000 Version 0.6.3 Released

* Testsuite case for lossy ground using somnec.

* somnec now works with g77 and other compilers which don't save
  local variables, i.e. use automatic variables by default.

* The -M and -m options now work with ground wave output from the RP
  card, i.e. RP 1 .....

## 10/10/1999 Version 0.6.2 Released
 
* Improved quality of error messages.

* Fixed truncation of output files with -M options.

* Improved man page for nec2d.


cnec2/nec.c
/*
 * aegnec2 - Dynamically Allocated Numerical Electromagnetics Code Version 2 
 * Copyright (C) 1998-2016 Ian David Flintoft <ian.flintoft@york.ac.uk>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * ---------------------------------------------------------------------
 *
 * nec.c -- C front end for NEC2D.
 *
 * This file contains the main program and options parsing routines of
 * the C front end.
 *
 */


#include <config.h>
 
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>
#include <error.h>
#include <getopt.h>

#include "cnec2d.h"
#include "nec.h"

/* 
 * Program call name. 
 */

char *program_name;


/*
 * Call NEC2D using the cnec2d interface with array
 * parameters determined by command line options. 
 */

int main( int argc, char **argv )
{

  char *nec_file = 0;                   /* Input file name */
  char *job_name = 0;                   /* Job name */
  int ret        = 0;                   /* Return code from cnec2d */
  int multi_file = MULTI_FILE_DEFAULT;  /* Multi file output flag */
  int verbose    = 0;                   /* Verbose output if true */
  int debug      = DEBUG_DEFAULT;       /* Debug level */
  int remove_old = 0;                   /* Remove old files if true */
  long num_seg   = NUM_SEG_DEFAULT;     /* Number of segments  */ 
  long num_patch = NUM_PATCH_DEFAULT;   /* Number of patches */
  long nsmax     = NSMAX_DEFAULT;       /* NSMAX */
  long loadmx    = LOADMX_DEFAULT;      /* LOADMX */
  long netmx     = NETMX_DEFAULT;       /* NETMX */
  long frqmax    = FRQMAX_DEFAULT;      /* Number of frequencies */
  long angmax    = ANGMAX_DEFAULT;      /* Number of angles */
  long mxcoup    = MXCOUP_DEFAULT;      /* MXCOUP */
  long jmax      = JMAX_DEFAULT;        /* JMAX */
  long nsmaxx    = NSMAXX_DEFAULT;      /* NSMAXX */
  long npmax     = NPMAX_DEFAULT;       /* NPMAX */
  long ld;                              /* LD */
  long iresrv;                          /* Core storage size multiple of 16 bytes */
  long normf;                           /* NORMF */
  double core_size = CORE_SIZE_DEFAULT; /* Core size in Mb */
  double d_num_seg;                     /* num_seg as a double */         
  double d_num_patch;                   /* num_seg as a double */         
  double d_ld;                          /* ld as a double */         
  double d_iresrv;                      /* Core storage size as a double */
  double in_core_iresrv;                /* IRESRV for in core solution */

  /* Set program name for error messages */
  program_name = argv[0];

  /* Get options and file name */
  get_options( argc, argv, &nec_file, &multi_file, &core_size, &num_seg, 
	       &num_patch, &nsmax, &loadmx, &netmx, &frqmax, &angmax, 
               &mxcoup, &jmax, &nsmaxx, &npmax, &verbose, &debug, &remove_old );
  
  /* Verify input file name and file exists */
  job_name = (char*) malloc( strlen( nec_file ) + 2 );
  strncpy( job_name, nec_file, strlen( nec_file ) + 1 );
  verify_input_file( &job_name );

  /* Remove old output files */
  if( remove_old )
    remove_old_files( job_name );

  /* FNORM must be larger of angmax and frqmax */
  if( angmax >= frqmax )
    normf = angmax;
  else
    normf = frqmax;
  
  /* LD depends on number of segments and patches */
  d_num_seg = (double)num_seg;
  d_num_patch = (double)num_patch;
  d_ld = d_num_seg + d_num_patch;

  if( d_ld <= (double)MAX_LONG ) 
    {
      if ( (long)d_ld > LD_MAX ) 
	{
	  free( job_name );
	  error( 1, 0, "LD exceeds hard limit of %d\n", LD_MAX );
	}
      else 
	{
	  ld = (long)d_ld;
	}
    }
  else 
    {
      free( job_name );
      error( 1, 0, "LD too big for long\n" );
    }
  
  /* Verfiy core size and set IRESRV */
  in_core_iresrv = ( d_num_seg + 2 * d_num_patch ) * ( d_num_seg + 2 * d_num_patch );
  
  if( core_size == 0 ) 
    {
      d_iresrv = in_core_iresrv;
      core_size = 16.0 * d_iresrv / (double)ONE_MEGABYTE;
    }
  else 
    {
      d_iresrv = core_size * (double)ONE_MEGABYTE / 16.0;
      if( d_iresrv < 2.0 * ( (double)num_seg + 2.0 * (double)num_patch ) ) 
	{
	  free( job_name );
	  error( 1, 0, "core size too small for job" );
	}
      else if ( d_iresrv < 0.2 * in_core_iresrv ) 
	{
	  error( 0, 0, "warning: possibly massively out of core solution" );
	}
      else if ( d_iresrv < in_core_iresrv ) 
	{
	  error( 0, 0, "warning: possibly out of core solution" );
	}
    }
  
  if( d_iresrv <= (double)MAX_LONG ) 
    {
      iresrv = (long)d_iresrv;
    }
  else 
    {
      free( job_name );
      error( 1, 0, "core size too big for long" );
    }
  
  /* Dump values if verbose */
  if( verbose )
    dump_values( nec_file, multi_file, core_size, num_seg, num_patch, nsmax, 
		 loadmx, netmx, frqmax, angmax, mxcoup, jmax, nsmaxx, 
		 npmax, ld, iresrv, normf );
  
  /* Call NEC */
  ret = cnec2d( job_name, ld, iresrv, nsmax, loadmx, netmx, 
                normf, mxcoup, jmax , nsmaxx, npmax, multi_file, 
		verbose, debug );
  
  if( ret != CNEC2D_ERR_OK ) 
    {
      free( job_name );
      error( 1, 0, "%s", cnec2d_err( ret ) );
    }
  
  free( job_name );
  
  return 0;

}


/*
 * Give version info and exit.
 */

void 
version_info( int verbose )
{
  
  printf( "\ncnec2 (%s) %s\n", PACKAGE, VERSION );
  printf( "Copyright (C) 1999-2002 Ian David Flintoft\n" );
  printf( "Maintained by Ian David Flintoft <idf1@ohm.york.ac.uk>\n" );
  printf( "%s comes with NO WARRANTY,\n", PACKAGE );
  printf( "to the extent permitted by law.\n" );
  printf( "You may redistribute copies of %s\n", PACKAGE );
  printf( "under the terms of the GNU General Public License.\n" );
  printf( "For more information about these matters,\n" );
  printf( "see the file named COPYING.\n\n" );

#ifdef USE_BLAS
  printf( "Using LAPACK/BLAS LU solver.\n\n" );
#else
  printf( "Using NEC2 built in LU solver.\n\n" );
#endif
#ifdef BUILTIN_SOMNEC
  printf( "SOMNEC is built in.\n\n" );
#else
  printf( "SOMNEC support requires external program.\n\n" );
#endif

  printf( "Built on %s at %s.\n\n", __DATE__, __TIME__ );

  if( verbose )
    {
      printf( "-------------------------------------------------\n");
      printf( "      Item              Minimum  Default  Maximum\n");
      printf( "-------------------------------------------------\n");
      printf( "Number of segments     %8d %8d %8d\n", 
	      NUM_SEG_MIN, NUM_SEG_DEFAULT, NUM_SEG_MAX );
      printf( "Number of patches      %8d %8d %8d\n", 
	      NUM_PATCH_MIN, NUM_PATCH_DEFAULT, NUM_PATCH_MAX );
      printf( "Core size (Mb)         %8d %8d %8d\n", 
	      CORE_SIZE_MIN, CORE_SIZE_DEFAULT, CORE_SIZE_MAX );
      printf( "Number of sources      %8d %8d %8d\n", 
	      NSMAX_MIN, NSMAX_DEFAULT, NSMAX_MAX );
      printf( "Number of loads        %8d %8d %8d\n", 
	      LOADMX_MIN, LOADMX_DEFAULT, LOADMX_MAX );
      printf( "Number of networks     %8d %8d %8d\n", 
	      NETMX_MIN, NETMX_DEFAULT, NETMX_MAX );
      printf( "Number of frequencies  %8d %8d %8d\n", 
	      FRQMAX_MIN, FRQMAX_DEFAULT, FRQMAX_MAX );
      printf( "Number of angles       %8d %8d %8d\n", 
	      ANGMAX_MIN, ANGMAX_DEFAULT, ANGMAX_MAX );
      printf( "Number of couplings    %8d %8d %8d\n", 
	      MXCOUP_MIN, MXCOUP_DEFAULT, MXCOUP_MAX );
      printf( "Number of connections  %8d %8d %8d\n", 
	      JMAX_MIN, JMAX_DEFAULT, JMAX_MAX );
      printf( "Number of NGF segments %8d %8d %8d\n", 
	      NSMAXX_MIN, NSMAXX_DEFAULT, NSMAXX_MAX );
      printf( "Number of NGF patches  %8d %8d %8d\n", 
	      NPMAX_MIN, NPMAX_DEFAULT, NPMAX_MAX );
      printf( "-------------------------------------------------\n\n");
    }

  exit( EXIT_SUCCESS );

  return;

}


/* 
 * Give usage info and exit. 
 */

void 
usage( const int exitcode, const char *error, const char *add )
{
  
  printf( "\ncnec2 (%s) %s\n", PACKAGE, VERSION );
  printf( "\n" );
  printf( "Usage:  %s [options] jobname[.nec]\n\n", PACKAGE );
  printf( "where jobname.nec is the name of a NEC input file. The \".nec\" suffix\n" );
  printf( "may be omitted. The valid options are:\n\n" );
  printf( "    -h, --help            Give this help and exit\n" );
  printf( "    -V, --version         Give version info and exit\n" );
  printf( "    -v, --verbose         Give verbose output\n" );
  printf( "    -d, --debug=N         Debug level (default 0)\n" );
  printf( "    -s, --segments=N      Maximum number of segments is N (default 300)\n" );
  printf( "    -p, --patches=M       Maximum number of patches is M (default 0)\n" );
  printf( "    -c, --core-size=N     Limit core size to N bytes. The suffices b, k\n" );
  printf( "                          and m may be used for bytes, kilobytes and megabytes\n" );
  printf( "                          respectivley (default to exactly fit number of segments\n" );
  printf( "                          and patches specified with -s and -p options)\n" );
  printf( "    -e, --sources=N       Maximum number of sources is N (default 30)\n" );
  printf( "    -l, --loads=N         Maximum number of loads is N ( default 30)\n" );
  printf( "    -n, --networks=N      Maximum number of networks is N (default 10)\n" );
  printf( "    -x, --connect=N       Maximum number of segments meeting at a single\n" );  
  printf( "                          wire junction is N (default 30)\n" );
  printf( "    -u, --coupling=N      Maximum number of points in coupling calculation\n" );
  printf( "                          is N (default 5)\n" );
  printf( "    -S, --ngf-segments=N  Maximum number of NGF segments to which new\n" );
  printf( "                          segments or patches connect is N (default 50)\n" );
  printf( "    -P, --ngf-patches=N   Maximum number of NGF patches to which new\n" );
  printf( "                          segments or patches connect is N (default 50)\n" );
  printf( "    -f, --frequencies=N   Maximum number of frequencies in normalised\n" );
  printf( "                          impedance calculation is N (default 200)\n" );
  printf( "    -a, --angles=N        Maximum number angles in received signal\n" );
  printf( "                          strength is N (default 200)\n" );
  printf( "    -m, --multi-file      Enable multiple file output\n" );
  printf( "    -M, --old-multi-file  Enable old style multiple file output\n" );
  printf( "    -r, --remove-old      Remove old output files before running solver\n" );
  printf( "\nReport bugs to Ian Flintoft <idf1@ohm.york.ac.uk>\n" ); 
  printf( "\n" );

  if(error) 
    fprintf( stderr, "%s: %s: %s\n", program_name, error, add );

  exit(exitcode);

  return;

}


/* 
 * Parse command line options, check bounds and determine
 * input file name.
 */

void 
get_options( int argc, char **argv , char **nec_file, int *multi_file, 
             double *core_size, long *num_seg, long *num_patch, 
             long *nsmax, long *loadmx, long *netmx, long *frqmax, 
	     long *angmax, long *mxcoup, long *jmax, long *nsmaxx, 
	     long *npmax, int *verbose, int *debug, int* remove_old )
{
  
  static struct option long_options[] = {
    { "version",           no_argument,       0, 'V' },
    { "help",              no_argument,       0, 'h' },
    { "verbose",           no_argument,       0, 'v' },
    { "debug",             required_argument, 0, 'd' },
    { "segments",          required_argument, 0, 's' },
    { "patches",           required_argument, 0, 'p' },
    { "core-size",         required_argument, 0, 'c' },
    { "sources",           required_argument, 0, 'e' },
    { "loads",             required_argument, 0, 'l' },
    { "networks",          required_argument, 0, 'n' },
    { "connect",           required_argument, 0, 'x' },
    { "coupling",          required_argument, 0, 'u' },
    { "ngf-segments",      required_argument, 0, 'S' },
    { "ngf-patches",       required_argument, 0, 'P' },
    { "frequencies",       required_argument, 0, 'f' },
    { "angles",            required_argument, 0, 'a' },
    { "multi-file",        no_argument,       0, 'm' },
    { "old-multi--file",   no_argument,       0, 'M' },
    { "remove-old",        no_argument,       0, 'r' },
    { 0, 0, 0, 0 }
  };
  char c;        /* Current short option */
  long tmp_long; /* Temporary long */

  while(1) 
    {
      
      c = getopt_long (argc, argv, "vhVd:s:p:c:e:l:n:x:u:S:P:f:a:mMr", long_options, 0 );
      
      if( c == -1 ) break;
      
      switch( c ) 
	{
	case 'V':
	  version_info( *verbose );
	  exit( EXIT_SUCCESS );
	  break;
	case 'h':
	  usage( EXIT_SUCCESS, 0, 0 );
	  break;
	case 'v':
	  *verbose = 1;
	  break;
	case 'd':
	  get_pos_long( c, &tmp_long, optarg, DEBUG_MIN, DEBUG_MAX );
	  *debug = tmp_long;
	  if( *debug > 0 ) 
	    *verbose = 1;
	  break;
	case 's':
	  get_pos_long( c, num_seg, optarg, NUM_SEG_MIN, NUM_SEG_MAX );
	  break;
	case 'p':
	  get_pos_long( c, num_patch, optarg, NUM_PATCH_MIN, NUM_PATCH_MAX );
	  break;
	case 'c':
	  get_pos_bkm( c, core_size, optarg, CORE_SIZE_MIN, CORE_SIZE_MAX );
	  break;
	case 'e':
	  get_pos_long( c, nsmax, optarg, NSMAX_MIN, NSMAX_MAX );
	  break;
	case 'l':
	  get_pos_long( c, loadmx, optarg, LOADMX_MIN, LOADMX_MAX );
	  break;
	case 'n':
	  get_pos_long( c, netmx, optarg, NETMX_MIN, NETMX_MAX );
	  break;
	case 'x':
	  get_pos_long( c, jmax, optarg, JMAX_MIN, JMAX_MAX );
	  break;
	case 'u':
	  get_pos_long( c, mxcoup, optarg, MXCOUP_MIN, MXCOUP_MAX );
	  break;
	case 'S':
	  get_pos_long( c, nsmaxx, optarg, NSMAXX_MIN, NSMAXX_MAX );
	  break;
	case 'P':
	  get_pos_long( c, npmax, optarg, NPMAX_MIN, NPMAX_MAX );
	  break;
	case 'f':
	  get_pos_long( c, frqmax, optarg, FRQMAX_MIN, FRQMAX_MAX );
	  break;
	case 'a':
	  get_pos_long( c, angmax, optarg, ANGMAX_MIN, ANGMAX_MAX );
          break;
	case 'm':
	  *multi_file = 1;
	  break;
	case 'M':
	  *multi_file = 2;
	  break;
	case 'r':
	  *remove_old = 1;
	  break;
	case '?':
	  error( 1, 0, "ambiguous match or an extraneous parameter" );
	  exit( EXIT_FAILURE );
	  break;
	case ':':
	  error( 1, 0, "missing parameter for option" );
	  break;
	default:
	  error( 1, 0, "unknown option %c", c );
      break;
	}
      
    }
  
  /* Get input file name as last arg */
  if( optind == argc )
    usage ( EXIT_FAILURE, "no input file name specified", "" );
  else if( optind == argc - 1 )
    *nec_file = argv[optind];
  else
    usage ( EXIT_FAILURE, "too many non-option arguments", "" );
  
  return;

}


/*
 * Parse a field consisting of a positive integer with an optional
 * suffix of b, k or m representing a sclae factor of 1, 1024 and
 * 1024*1024 respectively.
 *
 */

void
get_pos_bkm( const char c, double *val, const char *optarg, 
	     const long min, const long max )
{

  char *start;         /* Start of integer field */
  char *end;           /* End of integer field */
  char *junk;          /* Unused pointer in strtod */
  char *str;           /* Pointer to local copy of field */
  double scale = 1.0;  /* Scale factor from suffix */

  str = (char*) malloc( strlen( optarg ) + 1 );
  
  strncpy( str, optarg, strlen( optarg ) + 1 );

  start = str;

  if( *start == '+' ) start++;
  
  end = start;

  while( isdigit( *end ) ) end++;

  if( *end != '\0' ) 
    {
      switch( *end ) 
	{
	case 'b':
	  scale = 1.0;
	  *end = '\0';
      end++;
      break;
	case 'k':
	  scale = (double)ONE_KILOBYTE;
	  *end = '\0';
	  end++;
	  break;
	case 'm':
	  scale = (double)ONE_MEGABYTE;
	  *end = '\0';
	  end++;
	  break;
	default:
	  free( str );
	  error( 1, 0, "invalid value %s for option %c", optarg, c );
	  break;
	}
    }
  
  if( *end != '\0' ) 
    {
      free( str );
      error( 1, 0, "invalid value %s for option %c", optarg, c );
    }
  
  *val = strtod( start, &junk );
  *val *= scale;
  *val /= (double)ONE_MEGABYTE;
  
  if( *val < (long)min  || *val > (long)max ) 
    {
      free( str );
      error( 1, 0, "value of option %c must be in range %ldMb to %ldMb", 
	     c, min, max );
    }
  
  return;

}


/*
 * Parse a field consisting of a positive integer.
 *
 */

void
get_pos_long( const char c, long *val, char *optarg, 
	      const long min, const long max )
{

  char *start;    /* Start of integer field */
  char *end;      /* End of integer field */
  char *junk;     /* Unused pointer in strtod */
  double dval;    /* Value of field as a double */
  
  start = optarg;
  
  if( *start == '+' ) start++;
  
  end = start;
  
  while( isdigit( *end ) ) end++;
  
  if( *end != '\0' )
    error( 1, 0, "invalid value %s for option %c", optarg, c );
  
  dval = strtod( start, &junk );
  
  if( dval < (long)min  || dval > (long)max )
    error( 1, 0, "value of option %c must be in range %ld to %ld", c, min, max );
  else
    *val = (long)dval;
  
  return;

}


/*
 * Check input file exists and is openable.
 *
 * The name on the command line may have NEC_SUFFIX or not but
 * the actual file is required to have NEC_SUFFIX. Test for 
 * suffix on the file name - if not there add it.
 *
 */

void verify_input_file( char **job_name )
{

  FILE *fp;          /* Input file pointer */
  char *p;           /* Pointer to start of suffix */
  char *nec_file;    /* Pointer to file name */
  char *NEC_SUFFIX;  /* Input file suffix with dot */

  NEC_SUFFIX = (char *) malloc( strlen( SUFINP ) + 2 );
  strncpy( NEC_SUFFIX, ".", 2 );
  strncat( NEC_SUFFIX, SUFINP, strlen( SUFINP ) + 1 );
  p = strstr( *job_name, NEC_SUFFIX );

  if( p ) 
    {
      /* Suffix in file name */
      if( strncmp ( *job_name + strlen( *job_name ) - strlen( NEC_SUFFIX ), 
                    NEC_SUFFIX , strlen( NEC_SUFFIX ) + 1 ) != 0 ) 
	{
	  /* But not at end so ignore */
	  nec_file = (char *) malloc( strlen( *job_name ) + strlen( NEC_SUFFIX ) + 1 );
	  strncpy( nec_file, *job_name, strlen( *job_name ) + 1 );
	  strncat( nec_file, NEC_SUFFIX, strlen( NEC_SUFFIX ) + 1 );
	}
    else 
      {
	/* Strip off suffix for job name */
	nec_file = (char *) malloc( strlen( *job_name ) + 1 );
	strncpy( nec_file, *job_name, strlen( *job_name ) + 1 );
	*p = '\0';
      }
    }
  else 
    {
      /* No suffix - put one on to file name */
      nec_file = (char *) malloc( strlen( *job_name ) + strlen( NEC_SUFFIX ) + 1 );
      strncpy( nec_file, *job_name, strlen( *job_name ) + 1 );
      strncat( nec_file, NEC_SUFFIX, strlen( NEC_SUFFIX ) + 1 );
    }

  free( NEC_SUFFIX );

  /* Now try and open it */ 
  fp = fopen( nec_file, "r" );
  if( !fp ) 
    {
      free( nec_file );
      error( 1, 0, "cannot open input file %s", nec_file );    
    }

  if( ferror( fp ) )
    {
      free( nec_file );
      error( 1, 0, "error on input file %s", nec_file );    
    }

  if( fclose( fp ) == EOF )
    {
      free( nec_file );
      error( 1, 0, "cannot close input file %s", nec_file );    
    }
  
  return;
  
}


/* 
 * Dump out input values to stdout.
 *
 */

void 
dump_values( const char *nec_file, const int multi_file, const double core_size,
	     const long num_seg, const long num_patch, const long nsmax, 
	     const long loadmx, const long netmx, const long frqmax, 
	     const long angmax, const long mxcoup, const long jmax, 
	     const long nsmaxx, const long npmax, const long ld, 
	     const long iresrv, const long normf )
{

  printf( "\n\n" );
  printf( "Input file:             %s\n",  nec_file );
  printf( "Multi-file output:      %d\n",  multi_file );
  printf( "Core size (Mb):         %.1f\n", core_size );
  printf( "Number of segments:     %ld\n", num_seg );
  printf( "Number of patches:      %ld\n", num_patch );
  printf( "Number of sources:      %ld\n", nsmax );
  printf( "Number of loads:        %ld\n", loadmx );
  printf( "Number of networks:     %ld\n", netmx );
  printf( "Number of frequencies:  %ld\n", frqmax );
  printf( "Number of angles:       %ld\n", angmax );
  printf( "Value of NORMF:         %ld\n", normf );
  printf( "Number of couplings:    %ld\n", mxcoup );
  printf( "Number of connections:  %ld\n", jmax );
  printf( "Number of NGF segments: %ld\n", nsmaxx );
  printf( "Number of NGF patches:  %ld\n", npmax );
  printf( "Value of LD:            %ld\n", ld );
  printf( "Value of IRESRV:        %ld\n", iresrv );
  printf( "\n\n" );

  return;

}


/* 
 * Remove old output files.
 *
 */

void 
remove_old_files( const char *job_name )
{

  FILE *fpout;                     /* Output file pointer */
  char buf[10];                    /* Buffer for -m number */
  char *outfile  = 0;              /* Output file buffer */
  char suffices1[][5] = { ".res" , 
                          ".imp" , 
                          "" };    /* Suffices for -M files */
  char suffices2[][5] = { ".nre" , 
                          ".nrh" , 
                          ".rdp" , 
                          "" };    /* Suffice for -m files */
  int ret;                         /* Return code from remove */
  int i;                           /* Counter for suffix arrays */
  int j;                           /* Counter for -m files */
  int not_got_to_end;              /* True while more -m files left */
  
  outfile = (char *) malloc( strlen( job_name ) + CNEC2D_BUF_EXTRA );
  
  i=0;
  
  /* Delete files associated with -M option */
  
  while( suffices1[i][0] )
    {
      
      strncpy( outfile , job_name , strlen( job_name ) + 1 );
      strncat( outfile , suffices1[i] , strlen( suffices1[i] ) + 1 );  
      fpout = fopen( outfile , "r" );
      if ( fpout )
	{
	  fclose( fpout );
	  ret = remove( outfile );
	}
      
      i++;
    }
  
  i=0;
  
  while( suffices2[i][0] )
    {
      
      strncpy( outfile , job_name , strlen( job_name ) + 1 );
      strncat( outfile , suffices2[i] , strlen( suffices2[i] ) + 1 );  
      fpout = fopen( outfile , "r" );
      if ( fpout )
	{
	  fclose( fpout );
	  ret = remove( outfile );
	}
      
      i++;
    }
  
  i=0;
  
  /* Delete files associated with -m option */
  
  while( suffices2[i][0] )
    {
      
      j = 1;
      not_got_to_end = 1;
      
      do 
	{
	  
	  strncpy( outfile , job_name , strlen( job_name ) + 1 );
          sprintf( buf , "-%04d" , j );
	  strncat( outfile , buf , strlen( buf ) + 1 );  
	  strncat( outfile , suffices2[i] , strlen( suffices2[i] ) + 1 );  
	  fpout = fopen( outfile , "r" );
	  if ( fpout )
	    {
	      fclose( fpout );
	      ret = remove( outfile );
	    }
          else
	    not_got_to_end = 0;
	  
	  j++;
	  
	} while( not_got_to_end );
      
      i++;
    }
  
  return;
  
}
